File: 10-top-k-frequent/solution.java
// Top K Frequent Elements
// Time: O(n log k), Space: O(n) using min heap
// Alternative: O(n) using bucket sort

import java.util.*;

public class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        // Count frequencies
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }
        
        // Use min heap of size k
        PriorityQueue<Map.Entry<Integer, Integer>> minHeap = 
            new PriorityQueue<>((a, b) -> a.getValue() - b.getValue());
        
        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            minHeap.offer(entry);
            if (minHeap.size() > k) {
                minHeap.poll();
            }
        }
        
        // Extract results
        int[] result = new int[k];
        for (int i = 0; i < k; i++) {
            result[i] = minHeap.poll().getKey();
        }
        
        return result;
    }
    
    // Test
    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] result = sol.topKFrequent(new int[]{1, 1, 1, 2, 2, 3}, 2);
        System.out.println(Arrays.toString(result)); // Output: [1, 2] or [2, 1]
    }
}

File: 10-top-k-frequent/solution.ts
// Top K Frequent Elements
// Time: O(n log k), Space: O(n) using sorting
// Alternative: O(n) using bucket sort

class Solution {
    topKFrequent(nums: number[], k: number): number[] {
        // Count frequencies
        const freqMap = new Map<number, number>();
        for (const num of nums) {
            freqMap.set(num, (freqMap.get(num) || 0) + 1);
        }
        
        // Sort by frequency and get top k
        return Array.from(freqMap.entries())
            .sort((a, b) => b[1] - a[1])
            .slice(0, k)
            .map(entry => entry[0]);
    }
}

// Test
if (require.main === module) {
    const sol = new Solution();
    const result = sol.topKFrequent([1, 1, 1, 2, 2, 3], 2);
    console.log(result);  // Output: [1, 2]
}

export default Solution;
